File: 욕심쟁이 판다.cpp
#include<iostream>
#include<algorithm>
using namespace std;

int n,i,j;
int map[500][500];
int dp[500][500];
int dr[4] = { 1,-1,0,0 };
int dc[4] = { 0,0, 1,-1};
int solve(int row, int col) {
	int& ret = dp[row][col];
	if (ret != 0) return ret;
	ret = 1;
	for (int i = 0; i < 4; i++) {
		int nr = row + dr[i];
		int nc = col + dc[i];
		int num = map[row][col];
		
		if (nr < 0 || nc < 0 || n <= nr || n <= nc) continue;
		int n_num = map[nr][nc];
		if (n_num <= num) continue;
		dp[row][col] = max(dp[row][col], solve(nr,nc) + 1); 
	}
	return ret;
}
int main() {
	cin >> n;
	for (i = 0; i < n; i++) {
		for (j = 0; j < n; j++) {
			cin >> map[i][j];
		}
	}

	fill(&dp[0][0], &dp[0][0] + 250000, 0);
	
	int result = 0;
	for (i = 0; i < n; i++) {
		for (j = 0; j < n; j++) {
			if (dp[i][j] == 0) solve(i, j);;
			result = max(result, dp[i][j]);
		}
	}
	
	cout << result;
}

File: PS-DP/욕심쟁이 판다.py
import sys
sys.setrecursionlimit(10**5)
input = sys.stdin.readline

def dfs(row,col,cnt):
    dir = [[1,0],[-1,0],[0,1],[0,-1]]
    ret = cnt

    for i in range(4):
        newRow = row + dir[i][0]
        newCol = col + dir[i][1]
        if 0 <= newRow < n and 0 <= newCol < n:
            if forest[row][col] < forest[newRow][newCol]:
                if dp[newRow][newCol] != 0:
                    ret = max(ret, cnt + dp[newRow][newCol])
                else:
                    ret = max(ret, dfs(newRow,newCol,cnt + 1))

    dp[row][col] = ret - cnt + 1
    return ret

n = int(input())
dp = [[0] * n for _ in range(n)]
forest = []
result = 1
for _ in range(n):
    forest.append(list(map(int,input().split())))
for i in range(n):
    for j in range(n):
        if dp[i][j] == 0:
            dfs(i,j,1)
        result = max(result, dp[i][j])

print(result)